chain/client/src/chunk_executor_actor.rs

```diff
@@ -0,0 +1,533 @@
+use near_primitives::types::chunk_extra::ChunkExtra;
+use near_primitives::types::{AccountId, EpochId, ShardId, ShardIndex};
+use near_store::Store;
+use node_runtime::SignedValidPeriodTransactions;
+use tracing::instrument;
+
+pub struct ChunkExecutorActor {
+    chain_store: ChainStore,
+    runtime_adapter: Arc<dyn RuntimeAdapter>,
+    epoch_manager: Arc<dyn EpochManagerAdapter>,
+    /// Contains validator info about this node. This field is mutable and optional. Use with caution!
+    /// Lock the value of mutable validator signer for the duration of a request to ensure consistency.
+    /// Please note that the locked value should not be stored anywhere or passed through the thread boundary.
+    validator_signer: MutableValidatorSigner,
+    shard_tracker: ShardTracker,
+    network_adapter: PeerManagerAdapter,
+
+    /// Receipts originating from block keyed by block hash.
+    block_receipts_cache: LruCache<CryptoHash, HashMap<(ShardId, ShardId), ReceiptProof>>,
+}
+
+impl ChunkExecutorActor {
+    pub fn new(
+        store: Store,
+        genesis: &ChainGenesis,
+        runtime_adapter: Arc<dyn RuntimeAdapter>,
+        epoch_manager: Arc<dyn EpochManagerAdapter>,
+        validator_signer: MutableValidatorSigner,
+        shard_tracker: ShardTracker,
+        network_adapter: PeerManagerAdapter,
+        block_receipts_cache_capacity: NonZeroUsize,
+    ) -> Self {
+        Self {
+            chain_store: ChainStore::new(store, true, genesis.transaction_validity_period),
+            runtime_adapter,
+            epoch_manager,
+            validator_signer,
+            shard_tracker,
+            network_adapter,
+            block_receipts_cache: LruCache::new(block_receipts_cache_capacity),
+        }
+    }
+}
+
+impl near_async::messaging::Actor for ChunkExecutorActor {}
+
+/// Message with incoming receipts corresponding to the block.
+/// Eventually this would be handled properly with data availability layer.
+/// For now this is useful to do testing with test loop.
+#[derive(actix::Message, Debug)]
+#[rtype(result = "()")]
+pub struct ExecutorIncomingReceipt {
+    pub block_hash: CryptoHash,
+    pub receipt_proofs: Vec<ReceiptProof>,
+}
+
+/// Message that should be sent once block is processed to indicate that it's available for
+/// execution.
+#[derive(actix::Message, Debug)]
+#[rtype(result = "()")]
+pub struct ExecutorBlock {
+    pub block_hash: CryptoHash,
+}
+
+impl Handler<ExecutorIncomingReceipt> for ChunkExecutorActor {
+    fn handle(
+        &mut self,
+        ExecutorIncomingReceipt { block_hash, receipt_proofs }: ExecutorIncomingReceipt,
+    ) {
+        let block_receipts =
+            self.block_receipts_cache.get_or_insert_mut(block_hash, || HashMap::new());
+        for proof in receipt_proofs {
+            block_receipts.insert((proof.1.from_shard_id, proof.1.to_shard_id), proof);
+        }
+
+        let me = self.validator_signer.get().map(|signer| signer.validator_id().clone());
```

Comment:
What happens if there are two blocks after this block - we are on one fork right now and later switch to a different fork? The second fork wouldn't be executed at all if we no longer receive any new receipts for the previous block?